src/storage/DataAccessorBasedStore.ts

```diff
@@ -0,0 +1,366 @@
+    const isContainer = this.isNewContainer(representation.metadata, identifier.path);
+    if (oldMetadata && isContainer !== this.isExistingContainer(oldMetadata)) {
+      throw new ConflictHttpError('Input resource type does not match existing resource type.');
+    }
+    if (isContainer !== identifier.path.endsWith('/')) {
+      throw new UnsupportedHttpError('Containers should have a `/` at the end of their path, resources should not.');
+    }
+
+    // Potentially have to create containers if it didn't exist yet
+    await this.writeData(identifier, representation, isContainer, !oldMetadata);
+  }
+
+  public async modifyResource(): Promise<void> {
+    throw new Error('Not supported.');
+  }
+
+  public async deleteResource(identifier: ResourceIdentifier): Promise<void> {
+    this.validateIdentifier(identifier);
+    if (ensureTrailingSlash(identifier.path) === this.base) {
+      throw new MethodNotAllowedHttpError('Cannot delete root container.');
+    }
+    const metadata = await this.accessor.getMetadata(identifier);
+    if (metadata.getAll(LDP.contains).length > 0) {
+      throw new ConflictHttpError('Can only delete empty containers.');
+    }
+    return this.accessor.deleteResource(identifier);
+  }
+
+  /**
+   * Verify if the given identifier matches the stored base.
+   */
+  protected validateIdentifier(identifier: ResourceIdentifier): void {
+    if (!identifier.path.startsWith(this.base)) {
+      throw new NotFoundHttpError();
+    }
+  }
+
+  /**
+   * Returns the metadata matching the identifier, ignoring the presence of a trailing slash or not.
+   * This is used to support the following part of the spec:
+   * "If two URIs differ only in the trailing slash, and the server has associated a resource with one of them,
+   *  then the other URI MUST NOT correspond to another resource."
+   *
+   * First the identifier gets requested and if no result is found
+   * the identifier with differing trailing slash is requested.
+   * @param identifier - Identifier that needs to be checked.
+   */
+  protected async getNormalizedMetadata(identifier: ResourceIdentifier): Promise<RepresentationMetadata> {
+    const hasSlash = identifier.path.endsWith('/');
+    try {
+      return await this.accessor.getMetadata(identifier);
+    } catch (error: unknown) {
+      if (error instanceof NotFoundHttpError) {
+        return this.accessor.getMetadata(
+          { path: hasSlash ? trimTrailingSlashes(identifier.path) : ensureTrailingSlash(identifier.path) },
+        );
+      }
+      throw error;
+    }
+  }
+
+  /**
+   * Returns the result of `getNormalizedMetadata` or undefined if a 404 error is thrown.
+   */
+  protected async getSafeNormalizedMetadata(identifier: ResourceIdentifier):
+  Promise<RepresentationMetadata | undefined> {
+    try {
+      return await this.getNormalizedMetadata(identifier);
+    } catch (error: unknown) {
+      if (!(error instanceof NotFoundHttpError)) {
```

Comment:
I'm not sure how this case can happen, looking at `getNormalizedMetadata`, where `NotFoundHttpError` is explicitly caught and handled. But it's probably me, given test coverage.

Reply:
Only the first `NotFoundHttpError` is caught in `getNormalizedMetadata`. If that happens the other URI (with differing trailing slash) is accessed. That one can still throw.